crates/relayer/relayer-components/src/transaction/impls/message_sender/send_messages_with_indexed_signer.rs

```diff
@@ -0,0 +1,44 @@
+use alloc::format;
+use alloc::vec::Vec;
+
+use hermes_chain_components::traits::{HasChainId, MessageSenderComponent};
+use hermes_logging_components::traits::CanLog;
+use hermes_logging_components::types::LevelDebug;
+use hermes_prelude::*;
+
+use crate::chain::traits::MessageSender;
+use crate::transaction::traits::{CanSendMessagesWithSigner, HasMutexForSigner, HasSigner};
+
+pub struct SendMessagesWithIndexedSigner;
+
+#[cgp_provider(MessageSenderComponent)]
+impl<Chain> MessageSender<Chain> for SendMessagesWithIndexedSigner
+where
+    Chain:
+        HasChainId + HasSigner + HasMutexForSigner + CanSendMessagesWithSigner + CanLog<LevelDebug>,
+{
+    async fn send_messages(
+        chain: &Chain,
+        messages: Vec<Chain::Message>,
+    ) -> Result<Vec<Chain::MessageResponse>, Chain::Error> {
+        let (signer_index_mutex, max_signer_length) = chain.mutex_for_signer();
+        let mut signer_index = signer_index_mutex.lock().await;
+        let signer = chain.get_signer(*signer_index)?;
+        // Only update the signer index if there are additional signers
+        if max_signer_length > 0 {
+            // Since index 0 is used for the `key_entry` wallet, the maximum value for the index is 1 higher
+            // than the number of additional wallets.
+            *signer_index = (*signer_index + 1) % (max_signer_length + 1);
```

Comment:
IMO it would be better if we separate the round robin logic from here. we can come up with a generate struct.

```rs
pub struct RoundRobinIndex{
  last: usize,
  total: usize
}

impl RoundRobinIndex {
  fn new(total: usize) -> Self {
    Self { last: total - 1, total }
  }

  fn next(&mut self) -> usize {
    self.last = (self.last + 1) % total;
    self.last
  }
}

```

Reply:
Instead of adding an additional abstraction layer would it be ok to rename `SendMessagesWithIndexedSigner` to `SendMessagesWithRoundRobinSigner`?
I feel like the `MessageSender` already allows us to have different logic by implementing new components